django/test/runner.py

```diff
@@ -739,3 +742,25 @@ def setup_databases(verbosity, interactive, keepdb=False, debug_sql=False, paral
             connections[alias].force_debug_cursor = True
 
     return old_names
+
+
+def filter_tests_by_tags(suite, tags, exclude_tags):
+    suite_class = type(suite)
+    filtered_suite = suite_class()
+    for test in suite:
+        if isinstance(test, suite_class):
+            filtered_suite.addTests(filter_tests_by_tags(test, tags, exclude_tags))
+        else:
+            test_tags = set(getattr(test, 'tags', set()))
```

Comment:
I don't think the logic here is quite correct now, in terms of the interaction between `TestCase` level tags and test method level tags. The correct logic, I believe, is to first build up a set of all applicable tags (from both the `TestCase` and the test method), and then check that set of tags just once against `tags` and `exclude_tags`. The current logic here, if I'm reading it correctly, will short-circuit and accept a test if the `TestCase` level tags match, without ever checking the test-method tags, even if the test method is specifically tagged with an excluded tag.